proposed/local-sdk-global-json.md

```diff
@@ -0,0 +1,112 @@
+- `%VARIABLE%/$VARIABLE`: environment variables will be substituted. Either the Windows or Unix format can be used here and will be normalized for the operating system the host resolver executes on.
+
+This design requires us to only change the host resolver. That means other tooling like Visual Studio, VS Code, MSBuild, etc ... would transparently benefit from this change. Repositories could update global.json to have `additionalPaths` support `.dotnet` and Visual Studio would automatically find it without any design changes.
+
+## Considerations
+### Installation Points
+One item to keep in mind when considering this area is the .NET SDK can be installed in many locations. The most common are:
+
+- Machine wide
+- User wide: `%LocalAppData%\Microsoft\dotnet` on Windows and `$HOME/.dotnet` on Linux/macOS.
+- Repo: `.dotnet`
+
+Our installation tooling tends to avoid redundant installations. For example, if restoring a repository that requires 7.0.400, the tooling will not install it locally if 7.0.400 is installed machine wide. It also will not necessarily delete the local `.dotnet` folder or the user wide folder. That means developers end up with .NET SDK installs in all three locations but only the machine wide install has the correct ones.
+
+As a result solutions like "just use .dotnet, if it exists" fall short. It will work in a lot of casse but will fail in more complex scenarios. To completely close the disconnect here we need to consider all the possible locations.
+
+### Do we need additionalPathsOnly?
+The necessity of this property is questionable. The design includes it for completeness and understanding that the goal of some developers is complete isolation from machine state. As long as we're considering designs that embrace local deployment, it seemed sensible to extend the design to embrace _only_ local deployment.
+
+At the same time the motivation for this is much smaller. It would be reasonable to cut this from the design and consider it at a future time when the motivation is higher.
+
+### Best match or first match?
+This proposal is designed at giving global.json more control over how SDKs are found. If the global.json asked for a specific path to be considered and it has a matching SDK but a different SDK was chosen, that seems counter intuitive. Even in the case where the chosen SDK was _better_. This is a motivating scenario for CI where certainty around SDK is often more desirable than _better_. This is why the host discovery stops at first match vs. looking at all location and choosing the best match.
+
+Best match is a valid approach though. Can certainly see the argument for some customers wanting that. Feel like it cuts against the proposal a bit because it devalues `additionalPaths` a bit. If the resolver is switched to best match then feel like the need for `additionalPathsOnly` is much stronger. There would certainly be a customer segment that wanted to isolate from machine state in that case.
+
```

Comment:
Additional things worth mentioning:
* This will not affect the runtime resolution when executing `dotnet exec app.dll`
  * Maybe not a big deal - depends on the use cases - potentially we could introduce something like `dotnet exec --use-sdk-locations app.dll` or something like that to ask runtime resolution to also consider global.json

Reply:
* Compatibility considerations - this relies on the `dotnet.exe` which is on the user's path to be able to run the SDK specified in the `global.json`. This is not a big concern for the "Fallback to machine wide" since that will typically be the `dotnet.exe` used, but it is a concern for repo-local installs. This can lead to things like:
  * 8.0 `dotnet.exe` is used to run `9.0.0-preview...` SDKs (so we have to be really strict about forward compatibility)
  * 8.0 `dotnet.exe` is used to run `3.1.0` SDK (so we have to be really strict about backward compatibility)
None of this is exactly new, but specifically the forward compatibility doesn't have a common use case currently (it can happen, but it's very rare).

Reply:
The forward compatibility is somewhat concerning to me. Do we have a list of API dependencies that the muxer needs? And a strategy to ensure that those APIs don't change in the future?

Reply:
I could get you the list - it's VERY short (I think 2 APIs or maybe even just 1).
We are very aware of this requirement and we never touched those APIs. So I'm not super concern about this, but it does raise the importance of keeping it compatible since so far all the cases where it can happen are pretty rare, this would be basically the first real feature which makes use of it.

Reply:
Not only will `dotnet exec app.dll` not be affected, `dotnet app.dll` will also not be affected. This change will affect _SDK_ resolution only, not runtime resolution. `global.json` is not used for runtime execution.

Reply:
> This change will affect SDK resolution only, not runtime resolution.

I think this is they key part. But with potentially confusing aspects:
1. In order to run an SDK command, the host clearly needs to load same runtime. For this case, we should look for the runtime based on where the SDK was found. So in a sense it does affect runtime resolution, but only for running the SDK command, not for 'normal' app execution.
2. `dotnet exec` looks like an SDK command, but is not. It just runs the app and the SDK is not involved.
3. `dotnet run` is an SDK command that also runs the app. The SDK command launches the app, [setting the `DOTNET_ROOT` environment variable](https://github.com/dotnet/sdk/blob/9fd3b1eec66c9106433387749a3d490d8aec59f9/src/Cli/dotnet/commands/dotnet-run/RunCommand.cs#L261-L269) based on the running `dotnet`.
4. Other SDK commands [set `DOTNET_HOST_PATH`](https://github.com/dotnet/sdk/blob/9fd3b1eec66c9106433387749a3d490d8aec59f9/src/Cli/Microsoft.DotNet.Cli.Utils/MSBuildForwardingAppWithoutLogging.cs#L45) based on the running `dotnet`. Tools running as part of an SDK command look at its value.

For 3 and 4 in their current state, it would mean that the SDK command could be from some local deployment relative to global.json, but the command would then set variables such that child processes resolve the runtime based on the running `dotnet`, not where the SDK is running from. This may be odd? I can see the expectation that anything started due to one SDK command would be based on where that SDK was resolved. However, that would also mean `dotnet run` vs `dotnet app.dll` would be different in terms of where the runtime could come from.

Reply:
`dotnet run` and if I remember correctly also `dotnet test` actually have a somewhat more complex behavior. They will only propagate the SDK's runtime if the target is running on the same arch. They both need to support running apps/tests which target code incompatible with the SDK's runtime (for example running x86 target from x64 SDK). But for the same architecture they both try to make the target run on the same runtime as the SDK.

Reply:
The mechanics in `dotnet test` are different, but the overall goal was to achieve similar behavior to `dotnet run`.

Reply:
Correct we set VSTEST_WINAPPHOST_* vars and we use it in case of apphost usage(testhost.exe) https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs#L488

Before the `--arch` support we used the default naming `DOTNET_ROOT_*` but architecture switch uses these and so we had to "pass" the sdk root information in a different one.

Architecture switch uses the ``DOTNET_ROOT_*` ` to look for the correct muxer that can be set by users and can be different than the sdk one https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.CoreUtilities/Helpers/DotnetHostHelper.cs#L153

@nohwnd  something to add?

Reply:
> I'd like to see the "doesn't affect dotnet exec app.dll" part called out explicitly. Naively I would have expected it to, though I think it's ok that it doesn't.

I agree that tripped me up to so I will call it out. It materially impacts the scenarios here cause it is common in builds to effectively build a .NET core based tool then launch it with `dotnet exec`. That is done many times in the Roslyn build. 

At the same time it seems like we can continue this with a small tweak. Just need to flip to using `dotnet run` instead of `dotnet exec` in our scripts. That fixes other problems like having to hard code in the location of the built binary.

Reply:
> Other SDK commands [set DOTNET_HOST_PATH](https://github.com/dotnet/sdk/blob/9fd3b1eec66c9106433387749a3d490d8aec59f9/src/Cli/Microsoft.DotNet.Cli.Utils/MSBuildForwardingAppWithoutLogging.cs#L45) based on the running dotnet. Tools running as part of an SDK command look at its value.

Want to make sure that when we run `dotnet build` that `DOTNET_HOST_PATH` gets set to the _resolved_ runtime / SDK. Example it would get set to the app local .NET SDK. Pretty sure based on reading that is the case but wanted to double check.

Reply:
Want to elaborate on `DOTNET_HOST_PATH` a bit. Let's assume I have the following: 

- `dotnet.exe` is installed at `C:\Program Files\dotnet\dotnet.exe` and is on `%PATH%`. This location includes a 7.0.400 .NET SDK install
- The current directory is `c:\source` and has the following: 
  - `.dotnet` which is a directory that has 8.0.100 .NET SDK installed
  - `global.json` that requires `8.0.100` .NET SDK and has `path: ["$host$", ".dotnet" ]`. 

Now in the current directory I run `dotnet build` inside `c:\source`. That will eventually decide to use the 8.0.100 SDK from `c:\source\.dotnet` because the machine wide installation does not work.

What will the value of `%DOTNET_HOST_PATH%` be inside of msbuild here? That needs to be `c:\source\.dotnet` otherwise this proposal falls apart. The msbuild environment depends on this to `dotnet exec` tools inside of it so it must be a `dotnet.exe` that will have access to the 8.0.100 runtime. 

@elinor-fung, @agocke